docs: source: new file source tutorial (#585)
intel/dffml: docs/tutorials/sources/file.rst

(delimited by a string in square brackets) with each section containing one
or more name and value parameters.

Import modules
--------------

**dffml_source_ini/misc.py**

.. code-block:: python

    from configparser import ConfigParser

    from dffml import config, entrypoint, Record, FileSource, MemorySource, parser_helper

Here we are importing some common modules which will be required. The ``configparser``
module will be helpful in parsing INI files. The ``FileSource`` and ``MemorySource``
will be used as base class for our new Source. The config was imported to set the
configuration options for our new INISource. The entrypoint will be used to add the
entrypoint to our INISource. A ``Record`` is a unique entry in a source.

Add configuration
-----------------

**dffml_source_ini/misc.py**

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 11-15

Here we will be writing the configuration options which will be avaliable for our
INISource. The INISourceConfig will be decorated by config. Here we have provided
three configuration options.

Create Source
-------------

**dffml_source_ini/misc.py**

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 18-24

> Here we have added the entrypoint to INISource class as "ini". The new Source should inherit

Reviewer: We probably want to mention that there already is an `ini` source in dffml, so users will want to name theirs `myini` or something.

Author: sure

the data for that record.

Going over the code, we have defined a coroutine with parameter fileobj, here
fileobj is the file object. we are reading from the fileobj file object. Each section of the
INI file is used as a Record, with the name of the section used as key for that Record.
Each section consists of key value pairs stored as a dict. We're going to treat
this data as the feature data for each Record. To do so we pass the data as
the value for the ``features`` key under the ``data`` keyword argument when
creating a new Record.

Add dump method
---------------

**dffml_source_ini/misc.py**

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 49-66

This method will be used to dump the data to the file. We will read data from memory
(self.mem) and save that data in file object (fileobj).

Going over the code, we have defined a coroutine with parameter fileobj, here
fileobj is the file object. We are going over each section name and its corresponding Record.
We are reading all the data from the memory (self.mem) and writing that data to our file
object (fileobj). Hence dumping all our data into file.

Add Tests
---------

**tests/test_source.py**

> .. literalinclude:: ../../../tests/source/test_ini.py

Reviewer: ```suggestion
```

>     :lines: 1-7

Reviewer: ```suggestion
.. code-block:: python

    import os
    from tempfile import TemporaryDirectory

    from dffml import Record, load, save, AsyncTestCase

    from dffml_source_ini.misc import INISource
```

``load`` from high_level. save method will be used to save the records to the source and
load will be used to yield records from a source. AsyncTestCase will be used to run our
test methods as coroutines in default event loop.

**tests/test_source.py**

.. literalinclude:: ../../../tests/source/test_ini.py
    :lines: 10-30

To test the working of the INISource created we will create a class TestINISource, since
we are using the unittest testing framework. The TestINISource will inherit from AsyncTestCase
so that it can be run as a coroutine in the default event loop. In the test method we will
create a TemporaryDirectory which will contain our `.ini` file.

We will create an instance of our INISource with configuration options. Next we will use
the save method to save some records to the source. Do not forget to await the save method
as it is a coroutine. Next we will use the load method to yield records from the source.
load method will return AsyncIterator[Record] type object. Lastly, We need to check that
the records we saved is the same record which gets loaded.

Run the tests
-------------

To run the tests

.. code-block:: console

>     $ python setup.py test -s tests.test_source

Reviewer: ```suggestion
    $ python3 setup.py test -s tests.test_source
```

This will look into the file test_source.py and run all the tests.

Add the entrypoint
------------------

**setup.py**

To register your source under dffml entrypoint you need to add the following
to the setup.py file

.. code-block:: python

    common.KWARGS["entry_points"] = {

>         "dffml.source": [f"ini = {common.IMPORT_NAME}.misc:INISource"]

Reviewer: ```suggestion
        "dffml.source": [f"myini = {common.IMPORT_NAME}.misc:INISource"]
```

    }

This will add the newly created source to the dffml entrypoints and hence can
also be used in CLI.

Install your package
--------------------

To install your new source run

.. code-block:: console

    $ python3 -m pip install --prefix=~/.local -e .

CLI Usage
---------

To use your newly created source in CLI.

.. code-block:: console

    $ cat > data.ini << EOF
    [dffml]
    third_party = yes
    maintained = true

    [python]
    third_party = no
    maintained = true
    EOF

> 

Reviewer: ```suggestion
```

>     $ dffml list records -sources data=ini -source-filename data.ini

Reviewer: ```suggestion
    $ dffml list records -sources data=myini -source-filename data.ini
```

> .. _new_file_type_source_tutorial:

Reviewer: We don't need this reference because we can link to the top of this file if we want to using the `:doc:` directive. If you wanted to reference a section within a doc then you'd add a reference like this

```suggestion
```

Author: done, thanks

Source for New file types
=========================

This tutorial will help you with implementing your own ``Source`` for new file
types. You may want to do this if you have some specific file formats from which
you want to load and save data from. Here we will go through implementing a source
for ``.ini`` file formats.

About INI files
---------------

An INI file is a configuration file used by Operating Systems and programs
to initialize program settings. It contains sections for settings and preferences
(delimited by a string in square brackets) with each section containing one
or more name and value parameters.

Import modules
--------------

> .. literalinclude:: ../../../dffml/source/ini.py

Reviewer: ```suggestion
```

Author: done

>     :lines: 1-8

Reviewer: Recently we made it so that we can import everything `from dffml`. However, we can't do that from within DFFML itself.

Let's have this be a `.. code-block:: python` instead

```suggestion
.. code-block:: python

    from configparser import ConfigParser

    from dffml import config, entrypoint, Record, FileSource, MemorySource, parser_helper
```

Author: done

Here we are importing some common modules which will be required. The ``configparser``
module will be helpful in parsing INI files. The ``FileSource`` and ``MemorySource``
will be used as base class for our new Source. The config was imported to set the
configuration options for our new INISource. The entrypoint will be used to add the
entrypoint to our INISource. The ``Record`` is the dffml way of storing all the

> data of the file(s).

Reviewer: A Record is a unique entry in a source.

Author: done

Add configuration
-----------------

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 11-15

Here we will be writing the configuration options which will be avaliable for our
INISource. The INISourceConfig will be decorated by config. Here we have provided
three configuration options.

Create Source
-------------

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 18-24

Here we have added the entrypoint to INISource class as "ini". The INISource will inherit
from FileSource and MemorySource which provides read / write operations for file and

> memory. Set the CONFIG variable to the INISourceConfig which we created earlier. Next we

Reviewer: We should mention that setting the `CONFIG` variable is important because the instantiated version of `CONFIG` is accessible as `self.config`

Author: done, thanks

> will writing the load and dump methods for INISource.

Reviewer: It's important to say that we're inheriting from `FileSource` because that abstracts the saving and loading of files so that we only have to implement the `load_fd` and `dump_fd` methods. It takes care of decompression on load and re-compression on save if the files extension signifies that it's compressed.

We inherit from `MemorySource` because it implements the methods required by a
```
:py:class:`Source <dffml.source.source.BaseSourceContext>`
```

provided that `self.mem` contains `Record` objects.

Author: done

Add load method
---------------

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 26-47

This method will be used to load the data from the file(s). We will be reading data

> from the file object (ifile) and loading that data into memory (self.mem). Each Record

Reviewer: ```suggestion
from the file object (ifile) and loading that data into memory (self.mem). Each
:py:class:`Record <dffml.record.Record>`
```

Author: done

> instance consist of name(str type) of the record and data(dict type), with data having

Reviewer: ```suggestion
instance consist of name (str type) of the record and data (dict type), with data having
```

a key ``features`` which stores all the data for that record.

> Going over the code, we have defined a coroutine with two parameters self and ifile, here

Reviewer: We don't need to mention the `self` parameter, since most methods in Python take `self`

Author: done, thanks

ifile is the file object. we are reading from the ifile file object. Each section of the
INI file is used as a Record, with the name of the section used as key for that Record.
Each section consist of name and value pair which we store it as a dict, under that Record
features key.

Add dump method
---------------

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 49-66

This method will be used to dump the data to the file. We will read data from memory
(self.mem) and save that data in file object (fd).

> Going over the code, we have defined a coroutine with two parameters self and fd, here

Reviewer: Let's change `ifile` and `fd` to `fileobj`, that way we're consistent between methods.

Author: done renaming the ifile, fd to fileobj in docs

to initialize program settings. It contains sections for settings and preferences
(delimited by a string in square brackets) with each section containing one
or more name and value parameters.

Import modules
--------------

.. code-block:: python

    from configparser import ConfigParser

    from dffml import config, entrypoint, Record, FileSource, MemorySource, parser_helper

Here we are importing some common modules which will be required. The ``configparser``
module will be helpful in parsing INI files. The ``FileSource`` and ``MemorySource``
will be used as base class for our new Source. The config was imported to set the
configuration options for our new INISource. The entrypoint will be used to add the
entrypoint to our INISource. A ``Record`` is a unique entry in a source.

Add configuration
-----------------

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 11-15

Here we will be writing the configuration options which will be avaliable for our
INISource. The INISourceConfig will be decorated by config. Here we have provided
three configuration options.

Create Source
-------------

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 18-24

> Here we have added the entrypoint to INISource class as "ini". The New Source should inherit

Reviewer: ```suggestion
Here we have added the entrypoint to INISource class as "ini". The new Source should inherit
```

from FileSource and MemorySource as it abstracts the saving and loading of files so that
we only have to implement the load_fd and dump_fd methods. It takes care of decompression
on load and re-compression on save if the files extension signifies that it's compressed.
We inherit from MemorySource because it implements the methods required by a
:py:class:`Source <dffml.source.source.BaseSourceContext>` provided that self.mem contains

> Record objects.

Reviewer: ```suggestion
Record objects.

```

Set the CONFIG variable to the INISourceConfig which we created earlier. Setting
the CONFIG variable is important because the instantiated version of CONFIG is accessible

> as self.config Next we will writing the load and dump methods for INISource.

Reviewer: ```suggestion
as self.config.

Next we will writing the load and dump methods for INISource.
```

Add load method
---------------

.. literalinclude:: ../../../dffml/source/ini.py
    :lines: 26-47

This method will be used to load the data from the file(s). We will be reading data
from the file object (fileobj) and loading that data into memory (self.mem). Each
:py:class:`Record <dffml.record.Record>` instance consist of key (str type) of the
record and data (dict type), with data having a key ``features`` which stores all
the data for that record.

Going over the code, we have defined a coroutine with parameter fileobj, here
fileobj is the file object. we are reading from the fileobj file object. Each section of the
INI file is used as a Record, with the name of the section used as key for that Record.

> Each section consist of name and value pair which we store it as a dict, under that Record

Reviewer: ```suggestion
Each section consists of key value pairs stored as a dict. We're going to treat
this data as the feature data for each Record. To do so we pass the data as
the value for the ``features`` key under the ``data`` keyword argument when
creating a new Record.
```

> features key.

Reviewer: ```suggestion
```
